src/cider/exec/nextgen/operators/HashJoinNode.cpp

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include "exec/nextgen/operators/HashJoinNode.h"
+
+#include "exec/nextgen/context/CodegenContext.h"
+#include "exec/nextgen/jitlib/JITLib.h"
+#include "exec/nextgen/jitlib/base/ValueTypes.h"
+#include "type/plan/Expr.h"
+
+namespace cider::exec::nextgen::operators {
+using namespace jitlib;
+using namespace context;
+
+class BuildTableReader {
+ public:
+  BuildTableReader(utils::JITExprValue& buffer_values,
+                   ExprPtr& expr,
+                   JITValuePointer& index)
+      : buffer_values_(buffer_values), expr_(expr), index_(index) {}
+
+  void read() {
+    switch (expr_->get_type_info().get_type()) {
```

Comment:
Do we need to check type every time? Or we can get a type inferenced reader?

Reply:
One row and multiple columns in the build table are traversed each time here, and the expr of different columns requires different processing logic. The whole logic is in a loop, and this level of abstraction is consistent with the logic of ColumnToRow Translator‘s ColumnReader.

Reply:
I think it the codegen logic and once codegen done, no type check will need in row function right?

Reply:
Yes, the different column types of each row are consistent, but here are the different columns of a row, and the respective types may be different.